### ISSUE TITLE
DLPATH and make

### ISSUE THREAD

**Issue by videlec** [wbhart/flint2#447]
Compiling Sage with archlinux we are faced to a delicate problem:
 - make is linked to libguile (that link to libgc)
 - libgc is packaged by Sage with a different version and put in $SAGE_ROOT/local/lib
As a consequence, modifying `LD_LIBRARY_PATH` causes build failure (see [https://trac.sagemath.org/ticket/24575 upstream issue]). For flint, the culprit are the [lines 1-5 of Makefile.in](https://github.com/wbhart/flint2/blob/trunk/Makefile.in#L1-L5)
```
ifdef $(DLPATH)
  $(DLPATH):=$($(DLPATH)):$(DLPATH_ADD)
else
  $(DLPATH):=$(DLPATH_ADD)
endif
```
Removing them allow to build flint without troubles.

Note that we have a possible workaround in Sage using `LD_PRELOAD` (see the branch on [upstream ticket](https://trac.sagemath.org/ticket/24575)).

**Comment by videlec** [comment]
see also [arb issue 213](https://github.com/fredrik-johansson/arb/issues/213)

**Comment by wbhart** [comment]
Someone has added this patch to set DLPATH for some reason. It doesn't seem
that Flint itself actually uses it anywhere in the Makefile. So I'd need to
understand why it was added before I could just remove it. Reverting
someone else's fix isn't good practice in general.

I have no problem with changing it, but I need to understand what problem
it was originally fixing.

**Comment by wbhart** [comment]
JP Flori made the changes by the way. But it was just done as part of an
overhaul of the build system five years ago.

So the real question is, why should DLPATH ever be set. What is it actually
for, and why should it be set? Google doesn't seem to bring up any useful
information on it.

**Comment by videlec** [comment]
DLPATH is actually LD_LIBRARY_PATH (the change is visible in Makefile and not in Makefile.in).

**Comment by wbhart** [comment]
But if that is the case and I delete those lines, then surely make check won't work any more, since the library location won't be added to LD_LIBRARY_PATH.

**Comment by videlec** [comment]
Indeed. Would that be ok to set LD_LIBRARY_PATH only in the check part?

Concerning Sage, we are installing before testing with "make check". In such situation we could still use the workaround
```
make check DLPATH_ADD=
```

**Comment by wbhart** [comment]
I think that is the only reason it is set, so yes, it should be fine to set it only in the check part. Unfortunately I don't know how to do this. I don't know if you can stick an export directive inside a target like make check. The order in which make files are processed is frankly a mystery to me.

**Comment by wbhart** [comment]
Yeah, I don't think that is possible. Make doesn't specify an order of operation. A makefile consists of a bunch of targets and a bunch of variables, from what I can tell. The export directive just exports all the variables. It's not like a program that you get to direct with control flow.

At least there's no mention of anything like that in the documentation [1].

[1] https://www.lix.polytechnique.fr/~liberti/public/computing/prog/make.html

**Comment by videlec** [comment]
https://www.gnu.org/software/make/manual/html_node/Target_002dspecific.html

**Comment by wbhart** [comment]
But LD_LIBRARY_PATH is not just a variable here, it is exported to the environment. It's export that you need to be conditional, not the variable.

**Comment by wbhart** [comment]
Ok, but maybe they thought of that. We just set LD_LIBRARY_PATH to the modified value for make check, and then the export directive perhaps only exports that value if check is the current target.

The problem might be that check needs to build the library. So I'm not sure if it will manage that correctly or not.

**Comment by isuruf** [comment]
I think `LD_LIBRARY_PATH` and friends should be removed in favour of rpath. See my comment at https://github.com/wbhart/flint2/issues/321#issuecomment-280564955

**Comment by wbhart** [comment]
I guess that would work, since the test binaries are only intended to work in the source tree.

**Comment by wbhart** [comment]
Yeah, it looks like the RPATH option would fix a number of problems at once. So let's do it that way.

**Comment by videlec** [comment]
Should I provide a pull request?

**Comment by wbhart** [comment]
If you can do it, that would be great. I think Isuru's suggestion is probably the right way to go.

We have to be careful though. We can only set the RPATH for the test binaries, because libflint.so could be supplied by some binary service (e.g. we do this on Windows for Nemo.jl), and there's no way to know where it will end up being installed on the user's system. So only make check should be affected by the patch (except for the removal of the changes to LD_LIBRARY_PATH of course).

The RPATH should be set for the test binaries to the Flint source tree.

**Comment by wbhart** [comment]
One other thing we have to think about is MinGW-w64. I'm not sure how we are currently handling make check on that system. I don't think RPATH exists for Windows binaries, and LD_LIBRARY_PATH becomes PATH, I think.

**Comment by videlec** [comment]
This is now #450. I do not know how to fix/test on other platforms than mine... any help is welcome at this stage.

**Comment by wbhart** [comment]
Ah yeah, in MinGW and Cygwin, we set DLPATH=PATH instead of LD_LIBRARY_PATH and modify it. This behaviour has to remain the same for MinGW-w64, as there is no RPATH on MinGW-w64. It's probably ok to use RPATH for msys and Cygwin (though I didn't check this).

So this is a bit harder to fix than I thought.

**Comment by wbhart** [comment]
The other way to fix the problem for mingw-w64 is to use manifests.

**Comment by videlec** [comment]
(fixed with #450)
